samplekotlin/src/main/java/com/example/chattutorial/ChannelActivity4.kt

```diff
@@ -51,62 +51,51 @@ class ChannelActivity4 : AppCompatActivity(R.layout.activity_channel_4) {
                 is MessageListViewModel.Mode.Normal -> messageInputViewModel.resetThread()
             }
         }
-        // step 4 - let the message input know when we are editing a message
-        messageListView.setOnMessageEditHandler {
-            messageInputViewModel.editMessage.postValue(it)
-        }
 
-        // custom typing info header bar
-        val channelController = ChatClient.instance().channel(cid)
-        val currentlyTyping = MutableLiveData<Set<String>>(emptySet())
-
-        channelController.subscribeFor(TypingStartEvent::class.java, TypingStopEvent::class.java) {
-            val typing = currentlyTyping.value ?: emptySet()
-            val typingCopy = when (it) {
-                is TypingStartEvent -> {
-                    val name = it.user.extraData["name"] as String
-                    typing.toMutableSet() + name
-                }
-                is TypingStopEvent -> {
-                    val name = it.user.extraData["name"] as String
-                    typing.toMutableSet() - name
-                }
-                else -> {emptySet()}
+        messageListViewModel.state.observe(this) {
+            when (it) {
+                is MessageListViewModel.State.NavigateUp -> finish()
             }
-            currentlyTyping.postValue(typingCopy)
         }
 
-        val typingObserver = Observer<Set<String>> { users ->
-            channelHeaderSub.text = if (users.isEmpty()) {
-                "nobody is typing"
-            } else {
-                "typing: " + users.joinToString(", ")
-            }
+        // step 4 - let the message input know when we are editing a message
+        messageListView.setOnMessageEditHandler { message ->
+            messageInputViewModel.editMessage.postValue(message)
         }
-        currentlyTyping.observe(this, typingObserver)
+
+        val currentlyTyping = mutableSetOf<String>()
 
         // step 5 - handle back button behaviour correctly when you're in a thread
-        val backButtonHandler = {
+        channelHeaderView.onBackClick = {
             messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed)
         }
-        channelHeaderView.onBackClick = { backButtonHandler() }
 
-        onBackPressedDispatcher.addCallback(
-            this,
-            object : OnBackPressedCallback(true) {
-                override fun handleOnBackPressed() {
-                    backButtonHandler()
+        // make pressing back on hardware the same as channel header's back
+        onBackPressedDispatcher.addCallback(this) {
+            channelHeaderView.onBackClick()
+        }
+
+        // custom typing info header bar
+        ChatClient
+            .instance()
+            .channel(cid)
+            .subscribeFor(TypingStartEvent::class.java, TypingStopEvent::class.java) { event ->
```

Comment:
Was getting some weird behavior here, on my branch & current master. Not sure what's going on; will investigate more.

Reply:
Seems like if you start typing and pause for about 20 seconds, we stop getting events? Happens on `master` as well. Also, should a TypingStopEvent be triggered after a user hasn't entered more input after a certain period of time, or after message input is blank?

Reply:
Seems to have gone away now 🤷